README.txt
= zombies

* http://seattlerb.rubyforge.org/

== DESCRIPTION:

A toy zombie epidemic simulator using gosu.

== REQUIREMENTS:

* gosu

== LICENSE:

(The MIT License)

Copyright (c) Ryan Davis, seattle.rb

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 1.0.0 / 2010-09-22

* 1 major enhancement

  * Birthday!



test/test_zombies.rb
require "minitest/autorun"
require "minitest/benchmark" if ENV['BENCH']
require "zombies"
require "set"

class SpatialHash < Array
  attr_reader :width, :height, :unit
  def initialize width, height, unit
    @width, @height, @unit = width, height, unit

    @b = width  / unit

    cells = (height / unit) * (width / unit)

    super(cells) { Set.new }
  end

  def key x, y
    x = x.limit_to(0, width - 1)
    y = y.limit_to(0, height - 1)

    x = x.to_i / unit
    y = y.to_i / unit
    y * @b + x
  end

  def [] x, y = nil
    x, y = *x unless y
    raise ArgumentError, "need x,y or [x,y]" unless y
    at key(x, y)
  end
end

class TestZombies < MiniTest::Unit::TestCase
  attr_accessor :p

  def setup
    @p = Person.new nil, 37, 42
  end

  def test_tl
    assert_equal [32, 37], p.tl
  end

  def test_tr
    assert_equal [42, 37], p.tr
  end

  def test_bl
    assert_equal [32, 47], p.bl
  end

  def test_br
    assert_equal [42, 47], p.br
  end
end

class TestSpatialHash < MiniTest::Unit::TestCase
  attr_accessor :person, :h, :sw, :sh, :u

  def setup
    srand 0
    @sw, @sh, @u = 800, 600, 50
    @person = Person.new nil, @u-2, @u-2

    @h = SpatialHash.new @sw, @sh, @u
  end

  def test_prototype
    h[person.tl] << person
    h[person.tr] << person
    h[person.bl] << person
    h[person.br] << person

    assert_equal [person], h[person.tl].to_a
    assert_equal [person], h[person.tr].to_a
    assert_equal [person], h[person.bl].to_a
    assert_equal [person], h[person.br].to_a
  end

  def test_key
    assert_equal [43, 43], person.tl
    assert_equal [53, 43], person.tr
    assert_equal [43, 53], person.bl
    assert_equal [53, 53], person.br

    assert_equal   0, h.key(*person.tl)
    assert_equal   1, h.key(*person.tr)
    assert_equal  16, h.key(*person.bl)
    assert_equal  17, h.key(*person.br)
    assert_equal 191, h.key(801, 601)
  end

  def test_sanity
    srand 0

    @sw, @sh, @u = 40, 30, 10
    @h = SpatialHash.new @sw, @sh, @u

    people = (0...100).map { Person.new nil, rand(sw), rand(sh) }

    x = people.map { |op| [op.x, op.y] }

    assert_equal([[0, 3], [3, 7], [19, 21], [36, 23], [24, 24]],
                 x.first(5))

    people.each do |op|
      h[op.tl] << op
      h[op.tr] << op
      h[op.bl] << op
      h[op.br] << op
    end

    op        = h[0,0].sort_by { |oop| [oop.x, oop.y] }.first
    neighbors = h[op.tl] - [op] + h[op.tr] + h[op.bl] + h[op.br] - [op]
    visible   = neighbors.find_all { |oop| op - oop < op.sight }

    assert_equal  0, op.x
    assert_equal 3, op.y
    assert_equal 50, op.sight
    assert_equal 27, h[0,0].size # TODO: check this, seems high.
    assert_equal 26, neighbors.size
    assert_equal 26, visible.size

    new = people.map { |op|
      neighbors = h[op.tl] - [op] + h[op.tr] + h[op.bl] + h[op.br] - [op]
      visible   = neighbors.find_all { |oop| op - oop < op.sight }
      touching  = visible.find_all { |oop| op.touching? oop }
      touching.size
    }

    old = people.map { |op|
      visible  = people.find_all { |oop| op - oop < op.sight } - [op]
      touching = visible.find_all { |oop| op.touching? oop }
      touching.size
    }

    assert_equal old, new
  end

  attr_accessor :people

  def self.bench_range
    bench_linear 100, 1000, 100
  end

  def setup_bench n
    @people = (0...n).map { Person.new nil, rand(sw), rand(sh) }

    people.each do |op|
      h[op.tl] << op
      h[op.tr] << op
      h[op.bl] << op
      h[op.br] << op
    end
  end

  def bench_spatial_hash_old
    assert_performance_power 0.95 do |n|
      setup_bench n

      people.each do |op|
        visible  = people.find_all { |oop| op - oop < op.sight } - [op]
        touching = visible.find_all { |oop| op.touching? oop }
      end
    end
  end

  def bench_spatial_hash_new
    assert_performance_power 0.95 do |n|
      setup_bench n

      people.each do |op|
        neighbors = h[op.tl] - [op] + h[op.tr] + h[op.bl] + h[op.br] - [op]
        visible   = neighbors.find_all { |oop| op - oop < op.sight }
        touching  = visible.find_all { |oop| op.touching? oop }
      end
    end
  end
end


lib/zombies.rb
require 'gosu'

class Zombies < Gosu::Window
  VERSION = '1.0.0'

  if $DEBUG then
    W, H, FULL = 800, 600, false
  else
    W, H, FULL = Gosu::screen_width, Gosu::screen_height, true
  end

  # Tweakables:

  PEOPLE       = $DEBUG ? 100 : 200
  PRIESTS      = 5
  ZOMBIES      = 5
  BULLY_LIMIT  = 5
  PERSON_SIGHT = 50
  ZOMBIE_SIGHT = 40
  PRIEST_SIGHT = 100

  attr_reader :people
  attr_reader :humans
  attr_accessor :paused
  alias :paused? :paused

  def self.run n = nil
    window = Zombies.new n
    window.show
  end

  def initialize n = nil
    @t0 = Gosu.milliseconds
    @fps = FPSCounter.new
    @people, @humans = [], []
    @paused = false

    super W, H, FULL

    self.caption = "Zombie Epidemic Simulator"

    (n || PEOPLE).to_i.times do
      people << Person.new(self)
    end

    people[0...ZOMBIES].each do |person|
      person.infect!
    end

    people[ZOMBIES...ZOMBIES+PRIESTS].each do |person|
      person.fight!
    end

    humans.push(*people[ZOMBIES..-1])

    # randomize them so that priest/zombie battles are even
    people.replace people.sort_by { rand }
  end

  def priests myself = nil, &b
    result = people.grep Priest
    result.delete myself if myself
    result = result.find_all(&b) if b
    result
  end

  def zombies myself = nil, &b
    result = people.grep Zombie
    result.delete myself if myself
    result = result.find_all(&b) if b
    result
  end

  def calculate_fps
    return if paused?

    @fps.register_tick
    t1 = Gosu.milliseconds / 2000
    if t1 != @t0 then
      p @fps.fps
      @t0 = t1
    end
  end

  def update
    calculate_fps

    people.each do |p|
      p.update
    end unless paused?

    self.paused = true if humans.empty? or not people.any? { |p| Zombie === p }
  end

  def draw
    people.each do |person|
      person.draw
    end
  end

  def button_down id
    case id
    when Gosu::KbD, Gosu::KbH then
      $DEBUG = ! $DEBUG
    when Gosu::KbEscape, Gosu::KbX then
      close
    when Gosu::KbSpace then
      self.paused = ! self.paused
    when Gosu::MsLeft then
      if paused? then
        fake = Person.new(self, mouse_x, mouse_y)
        hits = people.find_all { |p| (fake - p).abs < fake.w }
        p hits
        p :people => hits.map { |p| people.index(p) }
        p :humans => hits.map { |p| humans.index(p) }
      end
    end
  end

  def needs_cursor?
    paused?
  end
end

class Person
  W, H, V, DA = 10, 10, 2, 30
  W2, H2, DA2 = W/2, H/2, DA/2

  attr_reader :window
  attr_accessor :x, :y, :v, :a, :neighbors, :panicked
  attr_accessor :chases
  alias :panicked? :panicked

  def initialize window, x = nil, y = nil
    @window = window
    @panicked = false
    @neighbors = []
    @chases = 0

    self.x = x || rand(window.width)
    self.y = y || rand(window.height)

    self.v, self.a = V, rand(360)
  end

  def sight
    Zombies::PERSON_SIGHT
  end

  def da;  DA;  end
  def da2; DA2; end
  def w;   W;   end
  def w2;  W2;  end
  def h2;  H2;  end

  def move
    dx, dy = Gosu.offset_xy a, v
    self.x = (x + dx).limit_to(0, window.width  - w2)
    self.y = (y + dy).limit_to(0, window.height - h2)
  end

  def update
    update_neighbors
    change_direction
    move
    attack
  end

  def update_neighbors
  end

  def change_direction
    if panicked? then
      panic
    else
      self.a += rand(da) - da2
      move_away_from_wall if near_edge?
    end
  end

  def attack
  end

  def hunt
    closest = neighbors.first
    self.a = Gosu::angle(x, y, closest.x, closest.y)
  end

  def move_away_from_wall
    self.a = Gosu::angle(x, y, window.width/2, window.height/2)
  end

  def panic
    self.a += rand(2*da) - da
    self.panicked -= 1
    self.calm! if self.panicked == 0
  end

  def draw
    window.draw_rect l, t, r, b, color
  end

  def draw_sight
    window.draw_circle x, y, sight, Gosu::Color::GRAY
  end

  def - o
    Gosu.distance x, y, o.x, o.y
  end

  def infect!
    window.humans.delete self
    extend Zombie
  end

  def fight!
    extend Priest
  end

  def panic!
    self.panicked = 10
    self.v = V * 2
  end

  def calm!
    self.panicked = false
    self.v = V
    self.chases += 1
    self.fight! if chases > Zombies::BULLY_LIMIT
  end

  def touching? o
    self - o < W
  end

  def r; x + W2; end
  def l; x - W2; end
  def t; y - H2; end
  def b; y + H2; end

  def tl; [l, t]; end
  def tr; [r, t]; end
  def bl; [l, b]; end
  def br; [r, b]; end

  def near_edge?
    ( l < w2 or
      t < h2 or
      r > window.width - w2 or
      b > window.height - h2 )
  end

  def color
    if panicked? then
      Gosu::Color::YELLOW
    else
      Gosu::Color::GRAY
    end
  end
end

module Zombie
  def draw
    super
    draw_sight if $DEBUG
  end

  def panic!
    # no brains... no panic
  end

  def color
    if neighbors.empty?
      Gosu::Color::DARK_GREEN
    else
      Gosu::Color::PURPLE
    end
  end

  def update_neighbors
    visible = window.humans.find_all { |op| self - op < op.sight }
    visible.each { |op| op.panic! } # FIX: this should be pull, not push

    self.neighbors.replace visible.find_all { |op|
      self - op < self.sight # think "smell"
    }.sort_by { |op| self - op }
  end

  def change_direction
    if neighbors.empty? then
      self.a += rand(da) - da2
    else
      hunt
    end
  end

  def attack
    neighbors.each do |op|
      break unless touching? op
      op.infect!
    end
  end

  def sight # more like smell
    Zombies::ZOMBIE_SIGHT
  end

  def v
    super / 2
  end

  def inspect
    super.gsub(/Person/, 'Zombie')
  end
end

module Priest
  def draw
    super
    draw_sight if $DEBUG
  end

  def color
    if neighbors.empty?
      Gosu::Color::WHITE
    else
      Gosu::Color::RED
    end
  end

  def sight
    Zombies::PRIEST_SIGHT
  end

  def v
    super * 1.25
  end

  def panic!
    # um... no
  end

  def infect!
    @hp ||= 50
    @hp -= 1
    super if @hp == 0
  end

  def update_neighbors
    self.neighbors.replace window.zombies { |op| self - op < self.sight }.
      sort_by { |op| self - op }
  end

  def change_direction
    if neighbors.empty? then
      self.a =
        if near_edge? then
          move_away_from_wall
        elsif op = window.priests(self) { |op| self - op < self.sight }.first
          Gosu::angle(op.x, op.y, x, y)
        else
          self.a + rand(da) - da2
        end
    else
      hunt
    end
  end

  def attack
    neighbors.each do |op|
      break unless touching? op
      window.people.delete op
    end
  end

  def inspect
    super.gsub(/Person/, 'Priest')
  end
end

class Numeric
  def limit_to min, max
    [min, [self, max].min].max
  end
end

module Gosu
  Color::PURPLE     = Gosu::Color.new 0xFF7A00BE
  Color::DARK_GREEN = Gosu::Color.new 0xff006600

  def self.offset_xy a, r
    return Gosu::offset_x(a, r), Gosu::offset_y(a, r)
  end

  class Window
    def draw_rect x1, y1, x2, y2, c
      draw_quad(x1, y1, c,
                x2, y1, c,
                x1, y2, c,
                x2, y2, c)
    end

    def draw_circle x, y, r, c
      xx, yy = Gosu.offset_xy 0, r

      (0..360).step(20) do |a|
        next if a == 0
        dx, dy = Gosu.offset_xy a, r
        draw_line x+xx, y+yy, c, x+dx, y+dy, c
        xx, yy = dx, dy
      end
    end
  end
end

class FPSCounter
  attr_reader :fps

  def initialize
    @current_second = Gosu::milliseconds / 1000
    @accum_fps = 0
    @fps = 0
  end

  def register_tick
    @accum_fps += 1
    current_second = Gosu::milliseconds / 1000
    if current_second != @current_second
      @current_second = current_second
      @fps = @accum_fps
      @accum_fps = 0
    end
  end
end
